torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -301,6 +301,93 @@ class SubgraphSlicer {
   AliasDb& aliasDb_;
   std::vector<Node*>& diff_nodes_;
 };
+
+c10::optional<bool> getProfileNodeRequiresGrad(Node* n) {
+  TORCH_INTERNAL_ASSERT(n->kind() == prim::profile);
+  if (!n->hasAttribute(attr::profiled_type)) {
+    return c10::nullopt;
+  }
+  auto& type = n->ty(attr::profiled_type);
+  if (type->castRaw<TensorType>() == nullptr) {
+    return c10::nullopt;
+  }
+  return type->expectRef<TensorType>().requiresGrad();
+}
+
+void AddRequiresGradToDifferentiableGraph(Node* diff_graph) {
+  TORCH_INTERNAL_ASSERT(diff_graph->kind() == prim::DifferentiableGraph);
+  const auto& subgraph = diff_graph->g(attr::Subgraph);
+  for (auto i : c10::irange(subgraph->outputs().size())) {
+    Value* output = subgraph->outputs()[i];
+    if (output->node()->kind() == prim::profile) {
+      // already have requires_grad info from this profile node
+      continue;
+    }
+    if (output->type()->castRaw<TensorType>() == nullptr) {
+      // non-tensors don't get profiled.
+      continue;
+    }
+    if (output->type()->expectRef<TensorType>().requiresGrad().has_value()) {
+      continue;
+    }
+
+    // this node doesn't have any requires_grad info.
+    // look at its uses to try to find a profile node.
+    std::vector<bool> requiresGradResults;
+    for (auto& use : diff_graph->output(i)->uses()) {
+      if (use.user->kind() == prim::profile) {
+        c10::optional<bool> req_grad_use;
+        if ((req_grad_use = getProfileNodeRequiresGrad(use.user)).has_value()) {
+          requiresGradResults.push_back(*req_grad_use);
+        }
+      }
+
+      // maybe the profile node got absorbed into a differentiable graph
+      if (use.user->kind() == prim::DifferentiableGraph) {
+        const auto& dg = use.user->g(attr::Subgraph);
+        // check all the uses of this graph input to look for profile nodes.
+        Value* dg_value = dg->inputs()[use.offset];
+        for (auto& dg_use : dg_value->uses()) {
+          if (dg_use.user->kind() == prim::profile) {
+            c10::optional<bool> req_grad_use;
+            if ((req_grad_use = getProfileNodeRequiresGrad(dg_use.user))
+                    .has_value()) {
+              requiresGradResults.push_back(*req_grad_use);
+            }
+          }
+        }
+      }
+    }
+
+    if (!requiresGradResults.empty()) {
+      bool requires_grad = std::any_of(
```

Comment:
@eellison does this seem like a safe fix to the nograd issue?

I can think of one case where the output *should* requires_grad but this won't do so: if all the uses of an diff. graph output that requires_grad are used in nograd sections, then all the profiled values will record requires_grad=False. But I think that's probably fine because then I don't think it matters that it requires_grad?